woocommerce/woocommerce-gutenberg-products-block
# Product gallery/add crop images

## What

Fixes https://github.com/woocommerce/woocommerce-blocks/issues/10632

This PR adds the image crop support when the crop images to fit setting is enabled.

## Why

Some merchants may want to display only square images uniformly so that we don't get any side effects from images that are different ratios which can cause the gallery slider to shift.

## Note

Since the "Crop images to fit" setting is disabled by default, my goal here is to minimize unnecessary processing as much as possible. So my approach here is not to register any sizes by default. Only when the "Crop images to fit" setting is enabled, then it would check if the image attachment had the cropped image size metadata and file and if it doesn't exists, it would generate it "on-the-fly".

The crop size dimension will be determined by the smallest dimension of the original image size.

## Testing Instructions



_Please consider any edge cases this change may have, and also other areas of the product this may impact._

1. Create a new product with a featured image that is a portrait (longer height than width). Optionally add some gallery images.
2. Go to Appearance->Editor->Templates->Single Product
3. Replace Product Image Gallery block with the Product Gallery Block
4. Enable "Crop images to fit" setting
5. Go to the product you created in step 1 in the frontend
6. Ensure the main product image you see is now cropped into a square ratio
7. Make sure the gallery images are also cropped into a square ratio when you click on any gallery image
8. When clicking on the main image to open up the dialog. Notice the image in the dialog is not cropped.
9. Go to Appearance->Editor->Patterns->Template Parts->Product Gallery
10. Enable the "Crop images to fit" setting
11. Now go back to the product you created in step 1 in the frontend
12. Click on the main image to open up the dialog. Now you should see a cropped image there as well.


## Screenshots or screencast



| Crop disabled | Crop enabled |
| ------ | ----- |
|![file.png](https://github.com/woocommerce/woocommerce-blocks/assets/2132595/59ad9757-2a15-44b4-be04-8bc13c018beb)        |![file.png](https://github.com/woocommerce/woocommerce-blocks/assets/2132595/945be3b3-435a-483a-bd6e-28d5bbbc5f07)       |

## WooCommerce Visibility


Required:


## Checklist

Required:

Conditional:

## Changelog


> Product Gallery: Add crop image feature
